csharp/src/Apache.Arrow/Arrays/PrimitiveArray.cs

```diff
@@ -86,5 +86,30 @@ IEnumerator IEnumerable.GetEnumerator()
                 yield return IsValid(index) ? Values[index] : null;
             }
         }
+
+        int ICollection<T?>.Count => Length;
+        bool ICollection<T?>.IsReadOnly => true;
+        void ICollection<T?>.Add(T? item) => throw new NotSupportedException("Collection is read-only.");
+        bool ICollection<T?>.Remove(T? item) => throw new NotSupportedException("Collection is read-only.");
+        void ICollection<T?>.Clear() => throw new NotSupportedException("Collection is read-only.");
+
+        bool ICollection<T?>.Contains(T? item)
```

Comment:
This could be probably be optimized as something like
```
            if (value == null)
            {
                return NullCount > 0;
            }
            else
            {
                ReadOnlySpan<T> values = Values;
                while (values.Length > 0)
                {
                    int index = Values.IndexOf(value.Value);
                    if (index < 0 || IsValid(index)) { return index >= 0; }
                    values = values.Slice(index + 1);
                }
                return false;
            }
```

Reply:
(The same applies to the other Index methods.)

Reply:
Using `ReadOnlySpan<T>.IndexOf` requires that the type `T` implements `IEquatable<T>`. Is this restriction desirable?
I tried to add this type constraint on line 24, but I could not get this constraint to coexist with the constraint that `T` be a `struct`. I might just have got the wrong syntax for this.

Reply:
I have implemented this type restriction, and used a (hopefully correct) simplified version of your code snippet above. Let's discuss whether this type restriction is desirable and worth it for the potential performance improvement (which I have not benchmarked).

Reply:
I think all the types where this matters should implement IEquatable<T>; the set of types supported by Arrow is effectively closed. That said, this is still potentially a breaking change if someone is using PrimitiveArray to implement a non-Arrow type but the workaround is fairly straightforward.